### ISSUE TITLE
[WIP] API Endpoints

### ISSUE THREAD

**Issue by joeyparis** [GeekMagicClock/HelloCubic-Lite#19]
# WIP - API Endpoints

It would be awesome if we could get an official list of API endpoints for reading settings as well as setting them. I've been able to reverse-engineer some of them via the network logs on the interface, but an official list would be great!

This is a work in progress. I believe I've gotten all of the access points used in the web client plus the `rot` for setting the rotation/orientation of the screen. 

Here is what I have compiled so far:

--

#### Example Request

```bash
CURL -X GET http://192.168.X.X/wifi.json
```

#### Example Response

```json
{
  "aps": [
    {
      "c": "1",
      "ss": "WiFi Network 1",
      "e": 1,
      "r": -79
    },
    {
      "c": "6",
      "ss": "WiFi Network 2",
      "e": 1,
      "r": -73
    },
    {
      "c": "6",
      "ss": "WiFi Network 3",
      "e": 1,
      "r": -79
    },
    {
      "c": "6",
      "ss": "WiFi Network 4",
      "e": 1,
      "r": -85
    }
  ]
}
```

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| s | SSID | string |
| p | Password | string |

#### Example Request

```bash
curl -X GET http://192.168.X.X/wifisave?s=NETWORK_SSID&p=NETWORK_PASSWORD
```

#### `GET /set?delay=`

How long does it take to start connecting to the WiFi after starting up:
It's useful when used in car and the WiFi of the car starts slowly, or used at home when the router is powered off and restarted at regular intervals.

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| delay | Wifi connection delay in seconds | integer |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?delay=10
```

#### `GET /set?cd1=&cd2=`

Location example, the capital city of Korea, please input: Seoul

[optional, but if there are several same city names in your country/region, you need to find the correct city code.]
**Tips：Where to find Your Own city code >>>**

[1.Click me to search your city code after internet is connected](https://openweathermap.org/find)

2.Input your city name, click search and then choose the result, open it, the number at the link address tail is your city code! eg: Seoul,KR https://openweathermap.org/city/1835848, 1835848 is city code of Seoul,KR.

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| cd1 | City code from [OpenWeatherMap.com](https://openweathermap.com) | string |
| cd2 | Unsure? Web client always sets to 1000. Could be secondary city | string |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?cd1=1835848&cd2=1000
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| w_u | Wind Speed Unit | string | `m/s`, `km/h`, `mile/h` |
| t_u | Temperature Unit | string | `ºF`, `ºC` |
| p_u | Atmospheric Pressure Unit | string | `hPa`, `kPa`, `mmHg`, `inHg` |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?w_u=m/s&t_u=%C2%B0C&p_u=hPa
```

#### `GET /set?w_i=`

By default, weather update interval is 20 minutes fixed, if you use your own api key, higher interval is allowed.

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| w_i | Weather update interval in minutes | integer |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?w_i=20
```

#### `GET /set?key=`

Professional option, if you are not clear, please do not modify this.
But set your own api key can get higher weather update rate and will be more stable compared using the public key.

Register your key [here](https://home.openweathermap.org/api_keys), tips: NEW key would take 1-2 hours to take effect, be patient.

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| key | Open Weather Map API key | string |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?key=OWM_API_KEY
```

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- |
| hc | Hour hex color code (w/ leading #) | string |
| mc | Minute hex color code (w/ leading #) | string |
| sc | Second hex color code (w/ leading #) | string |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?hc=%23ffffff&mc=%23feba01&sc=%23ff5900
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| hours | Enable or disable 12-hour time | integer | `0` for 24-hour format, `1` for 12-hour format |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?hour=1
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| colon | Enable or disable time-colon blink | integer | `0` for no blink, `1` for blink |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?colon=1
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| font | ID for desired font | integer | `1` for Default Big Font, `2` for Digital Font |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?font=1
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| day | Desired date format | integer | _Reference table below_ |

| `day` value | Format |
| ---- | ---- |
| 1 | DD/MM/YYYY |
| 2 | YYYY/MM/DD |
| 3 | MM/DD/YYYY |
| 4 | MM/DD |
| 5 | DD/MM |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?day=1
```

#### Parameters

| Name | Description | Type |
| --- | ---- | ---- | ---- |
| ntp | Domain of desired ntp server | url |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?ntp=pool.ntp.org
```

#### Parameters

| Name | Description | Type | Accepted Values |
| --- | ---- | ---- | ---- |
| dst | Enable or disable daylight savings time | integer | `0` for daylight standard time, `1` for daylight savings time |

#### Example Request

```bash
curl -X GET http://192.168.X.X/set?dst=1
```

---

## Scratch Pad
Everything below is my scratchpad of observed endpoints that I have not yet documented above. I'm just putting it here temporarily so that as I properly format everything, others can reference it to start working on their own projects.

```

# Set rotation
/set?rot=0
1
2
3
4 # default

# Set countdown timer
/set?yr=YYYY&mth=MM&day=DD

# Enable image auto display and interval
/set?i_i=5&autoplay=1

# Set theme
/set?theme=1 # Weather clock today
/set?theme=2 # Photo album
/set?theme=3 # Time style 1
/set?theme=4 # Time style 2
/set?theme=5 # Time style 3
/set?theme=6 # Days countdown
/set?theme=7 # Simple weather clock

# Set autotheme switch
/set?theme_list=0,0,0,0,0,0,0&sw_en=1&theme_interval=10 # Disable

# Set brightness
/set?brt=255

# Set nightmode
/set?t1=22&t2=6&b1=50&b2=73&en=1

# Reset to factory default
/set?reset=1 # Always returns a failed status even when successful

# Reboot
/set?reboot=1 # Always returns a failed status even when successful

# Get device storage
/spaces.json

# Set image
/set?img=%2Fimage%2F<IMAGE_NAME>

# Delete image
/delete?file=%2Fimage%2F<IMAGE_NAME>

# List images (as html)
/filelist?dir=/image

# Upload image
/doUpload?dir=/image
- Form data - file: binary

# Set weather clock gif
/set?gif=%2Fgif%2f<IMAGE_NAME>

# Upload gif
/doUpload?dir=/gif
- Form data - file: binary

# Status
/album.json
/app.json
/brt.json
/city.json
/colon.json
/config.json
/day.json
/daytimer.json
/delay.json
/dst.json
/font.json
/hour12.json
/key.json
/ntp.json
/space.json
/theme_list.json
/timebrt.json
/timecolor.json
/unit.json
/v.json
/w_i.json
/wifi.json

```

**Comment by the3dsandwich** [comment]
Hi, thanks for the information!
I just got mine a couple days ago and stumbled across your thread, may I ask what was your method reverse engineering it?
Maybe I can help discover something too

Also, off topic but how did you verify that rotation works? my cube replied FAIL each time I call via `curl -X GET 'http://<IP>/set?rot=2'`

Lastly, this is a ESP32-WROOM-32E right? Did you figure out how to update firmware if needed?

**Comment by MetalRain682237** [comment]
Thank you for this @joeyparis!
Any idea how to write custom text to the screen?
Otherwise my idea was to create jpeg images with the text on them and set it as the current image.

**Comment by dkgoodman** [comment]
I also wish I could write custom text to the screen. I'm in the middle of writing a python script that does almost everything the web page does, and I plan to add a function that will let me write text on an image or colored background and send it to the device. This way it could display things programmatically.

**Comment by MetalRain682237** [comment]
> I also wish I could write custom text to the screen. I'm in the middle of writing a python script that does almost everything the web page does, and I plan to add a function that will let me write text on an image or colored background and send it to the device. This way it could display things programmatically.

Yeah that's a good idea, I used JavaScript, but was able to write custom text into a gif or png and upload it

**Comment by qurle** [comment]
@joeyparis, thanks for the data you gathered.
I used in my own project for Small TV product by wrapping these endpoints into fully-typed functions.

Maybe it could help JS / TS guys, feel free to use [`tv-api.ts`](https://github.com/qurle/smalltv/blob/main/src/tv-api.ts)

<img width="631" height="202" alt="Image" src="https://github.com/user-attachments/assets/51721605-4185-4e74-8107-e380993e919b" />
